torch/_inductor/lowering.py

```diff
@@ -3912,20 +3926,115 @@ def fn(idx, return_index):
         else:
             return maxval
 
-    r1 = Pointwise.create(
+    out = Pointwise.create(
         device=x.get_device(),
         dtype=x.get_dtype(),
         inner_fn=functools.partial(fn, return_index=False),
         ranges=new_size,
     )
-    r2 = Pointwise.create(
+    offsets = Pointwise.create(
         device=x.get_device(),
-        dtype=torch.int64,
+        dtype=offset_dtype,
         inner_fn=functools.partial(fn, return_index=True),
         ranges=new_size,
     )
-    # TODO(jansel): should we force these to be realized?
-    return r1, r2
+    return out, offsets
+
+
+@register_lowering(prims._low_memory_max_pool2d_with_offsets, type_promotion_kind=None)
+def _low_memory_max_pool2d_with_offsets(
+    x, kernel_size, stride, padding, dilation, ceil_mode, *, offset_dtype
+):
+    # assert we are not on a fallback path, the inductor decomp should have guaranteed this
+    kernel_size, stride, padding, dilation, _ = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation, assert_fallback=False
+    )
+    return max_pool2d_with_offsets_impl(
+        x, kernel_size, stride, padding, ceil_mode=ceil_mode, offset_dtype=offset_dtype
+    )
+
+
+@register_lowering(aten.max_pool2d_with_indices, type_promotion_kind=None)
+def max_pool2d_with_indices(
+    x, kernel_size, stride=None, padding=0, dilation=1, ceil_mode=False
+):
+    kernel_size, stride, padding, dilation, use_fallback = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation
+    )
+    if use_fallback:
+        return fallback_max_pool2d_with_indices(
+            x, kernel_size, stride, padding, dilation, ceil_mode
+        )
+
+    out, offsets = max_pool2d_with_offsets_impl(
+        x, kernel_size, stride, padding, ceil_mode=ceil_mode, offset_dtype=torch.int32
+    )
+    indices = max_pool2d_offsets_to_indices_impl(
+        offsets, kernel_size[1], x.get_size()[-1], stride, padding
+    )
+    # If we landed here, the indices are not used for the backward pass, realize
+    # hint here so the conversion is fused into the forward loop.
+    indices.realize_hint()
```

Comment:
Realization in inductor is separate from the forward and backward partitioning. First we do aot_autograd partitioning, then we compile each separately.

I don't think we need to realize the indices here. If they're an output of forward saved for backward they will get realized as graph outputs.